(Win32 refresh) Threads_Win32

sukibaby opened the pull request:
This file handles OS side interaction as well as the creation of threads within the game -  std::unique_ptr and std::mutex are introduced here for their safety and memory management benefits.

Mostly removal of compatibility code targeting Windows XP and earlier. For example a custom wrapper `PortableSignalObjectAndWait` was removed in favor of directly calling `SignalObjectAndWait`.

The `TerminateThread` call in ThreadImpl_Win32::Halt has been replaced with SuspendThread since it's the proper way to do it in the Windows API. This should resolve a Windows specific issue where the game might start acting strangely indefinitely until the system is rebooted. Also  we'll use ExitProcess if the `Kill` flag is passed.

Don't bother with the try-catch in SetThreadName unless this is a debug build (since exceptions are disabled in ITGm, so it won't work as expected).

sukibaby committed:
(Win32 refresh) Threads_Win32

This file handles OS side interaction as well as the creation of threads within the game -  std::unique_ptr and std::mutex are introduced here for their safety and memory management benefits.

A custom wrapper PortableSignalObjectAndWait was removed in favor of directly calling SignalObjectAndWait.

In SetThreadName, we change from EXCEPTION_CONTINUE_EXECUTION to EXCEPTION_EXECUTE_HANDLER as suggested by the compiler to avoid a potential infinite loop.

ThreadImpl_Win32::Halt has been refactored to unequivocably call SuspendThread instead of TerminateThread, as TerminateThread doesn't allow for proper thread cleanup.

Don't bother with the try-catch in SetThreadName unless this is a debug build (since exceptions are disabled in ITGm, so it won't work as expected).

teejusb commented on src/arch/Threads/Threads_Win32.cpp:
```diff
@@ -9,16 +9,23 @@
 static void InitThreadIdMutex()
 {
-	if( g_pThreadIdMutex != nullptr )
+	if (g_ThreadIdMutexInitialized.load(std::memory_order_acquire))
```
You should be using an actual mutex (std::mutex) for this instead of an atomic variable. It's possible to initialize twice here.

Consider the following:

```cpp
atomic<bool> b{false};

void foo() {
  if (b.load()) {
    // ... do initialization stuff
    b.exchange(true, ...);
  }
}
```

If you start multiple threads then they can all get into the body of the condition because we haven't gotten to the statement that actually updates the bool. This is essentially the same scenario that you have above, even though I swapped the conditional to make it more clear.

teejusb replied:
Actually do we need the atomic bool at all? You can just check if the unique_ptr is initialized. It's bool operator determines whether an object is currently managed by the unique_ptr.

teejusb commented on src/arch/Threads/Threads_Win32.cpp:
```diff
@@ -9,16 +9,23 @@
+	if (g_ThreadIdMutexInitialized.load(std::memory_order_acquire))
+		return;
+
+	std::unique_ptr<MutexImpl_Win32> temp(new MutexImpl_Win32(nullptr));
```
Use `std::make_unique`

teejusb commented on src/arch/Threads/Threads_Win32.cpp:
```diff
@@ -197,41 +203,55 @@ static bool SimpleWaitForSingleObject( HANDLE h, DWORD ms )
 	switch( ret )
 	{
 	case WAIT_OBJECT_0:
+	{
```
It's actually fine to omit braces for switch statements unless we're declaring variables within them. Here and below.

teejusb commented:
Hmm it seems like we might want to call TerminateThread sometimes. The only time I can see Kill=true is basically right before we're about to crash the game (with some handling): https://github.com/itgmania/itgmania/blob/5d4f9dcb07493ed4c51d6be23e9b41978162305a/src/archutils/Win32/Crash.cpp#L352

Suspending seems like the wrong thing to do here since we don't plan to ever resume it.

sukibaby commented:
> Hmm it seems like we might want to call TerminateThread sometimes. The only time I can see Kill=true is basically right before we're about to crash the game (with some handling): https://github.com/itgmania/itgmania/blob/5d4f9dcb07493ed4c51d6be23e9b41978162305a/src/archutils/Win32/Crash.cpp#L352
> 
> Suspending seems like the wrong thing to do here since we don't plan to ever resume it.

Correct me if I'm wrong, but I think your analysis of the code here is correct _but_ this is the Windows API where things often don't make sense. And in this case we actually don't want to call TerminateThread in case it's absolutely a last resort. 

This is from the Microsoft docs at https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-terminatethread ...

![Screenshot_20250212_203513_Firefox Klar.jpg](https://github.com/user-attachments/assets/feedb80e-f1d1-4322-b7c6-ed532b4995a0)

The preferred way is to Suspend the thread and then close the handle in all cases except perhaps in a crash handler.

sukibaby commented on src/arch/Threads/Threads_Win32.cpp:
```diff
@@ -6,37 +6,46 @@
+void ThreadImpl_Win32::Halt(bool /*Kill*/)
 {
-	if( Kill )
-		TerminateThread( ThreadHandle, 0 );
```
I think it would be better to use `ExitThread` or `ExitProcess` here if you prefer to keep the Kill booleans's functionality. 

https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-exitprocess?redirectedfrom=MSDN

sukibaby committed:
use ExitThread instead of TerminateThread

Preferred in Windows, especially when we can't guarantee that we are able to perform a proper clean-up before terminating. Halt may be called with the Kill flag which indicates everything must shut down. This is the preferred approach in the Windows API.

Another preferred option is to use WaitForSingleObject with a timeout, but Halt with the Kill flag is being called when the game is about to crash, so that's not a realistic option and may cause the program to hang indefinitely.

sukibaby committed:
undo whitespace changes in Threads_Win32
